### PR TITLE
fix: MLFlowLogger artifact_path

### LINKED ISSUE #20664: MLFlowLogger fails to log artifact on Windows
### Bug description

Error when training with "MLFlowLogger" and with `log_models="all"`, running on Windows:

```
mlflow.exceptions.MlflowException: Invalid artifact path: 'epoch=0-step=43654'. Names may be treated as files in certain cases, and must not resolve to other names when treated as such. This name would resolve to 'epoch=0-step=43654'.
```

I was able to find the reason for this error:

`MLFlowLogger` is calling `MLflowClient.log_artifact(...)` internally -- inside method `MLFlowLogger._scan_and_log_checkpoints(...)` -- passing the artifact path as a `pathlib.Path` object. However, MLFlow expects paths in the POSIX format, whereas `pathlib.Path` will use the current filesystem format.

Now, the actual reason for the error is due to an internal check that `MLFlow` does. It tries to verify that the path is already "normalized", that is, doesn't contain "." or "..". The way it does that is by calling `posixpath.normpath(...)` and checking if the output is the same as the original path. Because the original path is a "pathlib.Path" object, both paths do not match. I am assuming that if the platform was different from Windows, the comparison would match because the internal string representation of `pathlib.Path` would be the same.

### What version are you seeing the problem on?

v2.5

### How to reproduce the bug

```python

```

### Environment

<details>
  <summary>Current environment</summary>

```
#- PyTorch Lightning Version (e.g., 2.5.0): 2.5.1
#- PyTorch Version (e.g., 2.5): 2.6.0
#- Python version (e.g., 3.12): 3.12.9
#- OS (e.g., Linux): Windows
#- CUDA/cuDNN version: 12.6
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source): pip
```

</details>

### More info

_No response_

### PR DESCRIPTION
**Pull request by yxtay**
## What does this PR do?

#20538
- It looks like the changes introduced in this PR broke mlflow logging of checkpoints
- `MlFlowClient.log_artifact` takes string for the `artifact_path` argument only
- I am getting the following error as a result. Not exactly helpful, but I figured the type difference was the reason.
  `mlflow.exceptions.MlflowException: Invalid artifact path: 'epoch=0-step=1'. Names may be treated as files in certain cases, and must not resolve to other names when treated as such. This name would resolve to 'epoch=0-step=1'`
- This change fixes it.

Fixes #20664

```python
# example.py

import os

import lightning as L
import torch
import torch.nn.functional as F
from torch import nn
from torch.utils.data import DataLoader
from torchvision import transforms
from torchvision.datasets import MNIST


class Encoder(nn.Module):
    def __init__(self):
        super().__init__()
        self.l1 = nn.Sequential(nn.Linear(28 * 28, 64), nn.ReLU(), nn.Linear(64, 3))

    def forward(self, x):
        return self.l1(x)


class Decoder(nn.Module):
    def __init__(self):
        super().__init__()
        self.l1 = nn.Sequential(nn.Linear(3, 64), nn.ReLU(), nn.Linear(64, 28 * 28))

    def forward(self, x):
        return self.l1(x)


class LitAutoEncoder(L.LightningModule):
    def __init__(self, encoder, decoder):
        super().__init__()
        self.encoder = encoder
        self.decoder = decoder

    def training_step(self, batch, batch_idx):
        # training_step defines the train loop.
        x, _ = batch
        x = x.view(x.size(0), -1)
        z = self.encoder(x)
        x_hat = self.decoder(z)
        loss = F.mse_loss(x_hat, x)
        return loss

    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=1e-3)
        return optimizer


dataset = MNIST(os.getcwd(), transform=transforms.ToTensor())
train_loader = DataLoader(dataset)
# model
autoencoder = LitAutoEncoder(Encoder(), Decoder())

# train model
mlflow_logger = L.pytorch.loggers.mlflow.MLFlowLogger(log_model=True)
trainer = L.Trainer(logger=mlflow_logger, max_steps=50)
trainer.fit(model=autoencoder, train_dataloaders=train_loader)
```

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--20669.org.readthedocs.build/en/20669/

### CODE REVIEW

File: src/lightning/pytorch/loggers/mlflow.py [hunk 1]
```diff
@@ -363,7 +363,7 @@ def _scan_and_log_checkpoints(self, checkpoint_callback: ModelCheckpoint) -> Non
             aliases = ["latest", "best"] if p == checkpoint_callback.best_model_path else ["latest"]
 
             # Artifact path on mlflow
-            artifact_path = Path(self._checkpoint_path_prefix) / Path(p).stem
+            artifact_path = str(Path(self._checkpoint_path_prefix, Path(p).stem))
```
Comment (niander) [inline src/lightning/pytorch/loggers/mlflow.py:366]: Instead of doing this string conversion, why not call the `as_posix()` function from the `Path` object. This should make this working on Windows as well. Please check this bug I filed: #20664 

`MLFlowClient` implementation expects paths in the POSIX format.

**Review by niander — CHANGES_REQUESTED** [review]
This fix should also make sure that this change will work on Windows when also passing a `checkpoint_path_prefix` to the `MLFlowLogger` object creation. `str(...)` on a `pathlib.Path` created on Windows may be translated into backward slashes which doesn't seem to be what `MLFlowClient` expects

File: src/lightning/pytorch/loggers/mlflow.py (hunk 1 above)
Comment (yxtay) [reply src/lightning/pytorch/loggers/mlflow.py:366]:
```suggestion
            artifact_path = Path(self._checkpoint_path_prefix, Path(p).stem).as_posix()
```

Comment (yxtay) [reply src/lightning/pytorch/loggers/mlflow.py]: Based on the comment on the issue, does it mean that it is necessary to call `resolve()` on the `artifact_path`?

Comment (niander) [reply src/lightning/pytorch/loggers/mlflow.py]: I believe `resolve()` is for resolving symlinks and special path components. `artifact_path` is holds the path for the artifact in the `MLFlow` tracking system, not the real local filesystem path. I would argue that using `as_posix` is enough and there is no need to use `resolve()`.

**Review by niander — APPROVED** [review]

**Comment by Northo** [comment]
Great to see this fixed! Do you know when we will have a release with this patch?

**Comment by niander** [comment]
@Borda can we get this into a new release soon? What is the release cadence? This fix unblocks people using MLFlowLogger for checkpoints.

**Comment by MohammadElSakka** [comment]
I'm still getting the same problem.
Any help would be appreciated :-) 
```
torch                              2.6.0+cu118
torchaudio                        2.6.0+cu118
torchmetrics                       1.7.1
torchvision                        0.21.0+cu118
lightning                          2.5.1.post0
lightning-utilities              0.14.3
pytorch-lightning             2.5.1.post0
mlflow                             2.22.0
mlflow-skinny                  2.22.0
```

**Comment by yxtay** [comment]
The PR changes were reverted in the master branch and not included in the `2.5.1.post0`. Not sure what the reason was, just stating what I found.

f6ef4099b6996e03fe345cfb5f9f2f5b3cad159b

**Comment by niander** [comment]
Also, I am very confused on why they reverted. It seems though that the they reverted more than just this but also the other commit that added custom path prefix for the artifacts.
@Borda

**Comment by Clara-liu** [comment]
Our CI is failing which caused us having to pin lightning to older version. Any reason this fix was reverted? Thanks.

**Comment by MohammadElSakka** [comment]
Is there an older version that is stable that has this fixed? Which workaround are you guys doing? It is a bit frustrating to allocate a whole section in my README.md to explain how to fix MLFlow in my program ... :melting_face:

**Comment by Clara-liu** [comment]
> Is there an older version that is stable that has this fixed? Which workaround are you guys doing? It is a bit frustrating to allocate a whole section in my README.md to explain how to fix MLFlow in my program ... 🫠

We've pinned lightning to 2.5.0 until this is resolved.

**Comment by MohammadElSakka** [comment]
> > Is there an older version that is stable that has this fixed? Which workaround are you guys doing? It is a bit frustrating to allocate a whole section in my README.md to explain how to fix MLFlow in my program ... 🫠
> 
> We've pinned lightning to 2.5.0 until this is resolved.

I'll experiment with that :saluting_face:

**Comment by niander** [comment]
@Borda could you please give some clarity on why this was reverted?

**Comment by aloysius-lim** [comment]
Hi @Borda any updates on this please? It has been blocking us from upgrading Lightning for a few months now.

**Comment by andremoeller** [comment]
@Borda +1 to this -- the revert is blocking our upgrade, so we're pinned to 2.5.0.

**Comment by patrontheo** [comment]
@Borda same issue here

**Comment by Borda** [comment]
@aloysius-lim @andremoeller @patrontheo, you need to talk to Will; it was not his decision :)

**Comment by patrontheo** [comment]
@Borda I'm not sure to undertand, whose decision was it then ?
And who is Will ?

**Comment by patrontheo** [comment]
Thanks :)
Then I guess @williamFalcon, could you please give some clarity on why this was reverted?

**Comment by joncarter1** [comment]
+1, I don't understand why this was reverted and the original feature added by @benglewis in https://github.com/Lightning-AI/pytorch-lightning/pull/20538 removed (@Borda, @lantiga)

The root cause of this issue was fixed in MLFlow 3.6.0, so the MLFlowLogger should now work for lightning versions up to 2.5.6, which includes the useful `checkpoint_path_prefix` feature. (FYI @benglewis, @niander, @Northo, @Clara-liu, @aloysius-lim, @andremoeller, @patrontheo, @MohammadElSakka )
